### PR TITLE
Migrate Cohere ONNX export to dynamo (torch.export) path

### LINKED ISSUE #4: Migrate Cohere ONNX export off legacy TorchScript exporter
## Summary
`public/scripts/cohere_export/export_cohere_transcribe_to_onnx.py` currently uses the legacy TorchScript-based `torch.onnx.export(..., dynamo=False)` path.

PyTorch now emits a deprecation warning during export:

> You are using the legacy TorchScript-based ONNX export. Starting in PyTorch 2.9, the new torch.export-based ONNX exporter has become the default.

## Why this matters
- The current export still works, but it is depending on a deprecated path.
- Future PyTorch releases may make this path harder to maintain or eventually remove it.
- The Cohere exporter has several export-time monkey patches, so we should validate compatibility with the new exporter deliberately rather than waiting for a forced migration.

## Scope
- Evaluate `torch.export` / modern ONNX exporter compatibility for:
  - `mel.onnx`
  - `encoder.onnx`
  - `decoder_init.onnx`
  - `decoder_step.onnx`
  - optional `decoder.onnx`
- Preserve current parity expectations and dynamic axes behavior.
- Reassess which monkey patches are still needed under the new exporter.

## Acceptance criteria
- At least one migration plan is documented, with risks called out.
- Ideally, the Cohere exporter no longer emits the legacy exporter deprecation warning.
- Parity/validation remains comparable to the current exporter.

## Notes
Current export validations were good in the latest local test:
- encoder parity max diff ~ `5.83e-4`
- mel parity max diff ~ `3.6e-5`
- KV parity max logit diffs ~ `3.78e-3` and `3.18e-2`

This should be treated as a follow-up migration task, not an emergency fix.

### PR DESCRIPTION
**Pull request by christopherthompson81**
## Summary
- Default Cohere ONNX export now uses `torch.onnx.export(dynamo=True)` with `dynamic_shapes` (`torch.export.Dim`) per call site, eliminating the legacy TorchScript exporter deprecation warning under PyTorch ≥ 2.9.
- Legacy path remains available behind `--legacy-exporter` for parity comparison and troubleshooting.
- Existing encoder monkey patches (RelPositionMultiHeadAttention forward, `_needs_conv_split`, `_materialize_pe`) are kept — `torch.export` is *stricter* about Python-bool branches on dynamic shapes, so removing them up front is still the right move.

## Notable mechanics
- `_run_torch_export()` is the unified entry point that selects legacy vs modern based on the new flag and converts `dynamic_axes` ↔ `dynamic_shapes` per call.
- Opset auto-clamps to ≥ 18 on the modern path because the dynamo registry only ships v18+ implementations for some ops (`Pad` lacks a v17 adapter); legacy path still accepts 17. Default opset bumped from 17 → 18.
- KV-cache wrappers use `B=2` / `past_len=2` dummy shapes so `torch.export` does not specialise the symbolic dim to `1`. The exported graphs still accept seq_len=1 at runtime.
- `_consolidate_external_data` now detects the case where the dynamo exporter has already written a single consolidated `<name>.onnx.data` file and skips the resave (it still sweeps stale `Constant_*` files).
- Adds `onnxscript >= 0.6` runtime dep; bumps `torch` floor to `>= 2.9`.

## Parity (CPU, fp32, vs legacy reference numbers from issue #4)
| Submodel       | Legacy max-diff | Modern max-diff |
| -------------- | --------------- | --------------- |
| `mel`          | ~3.6e-5         | ~3.5e-5         |
| `encoder`      | ~5.83e-4        | ~3e-6           |
| `decoder_init` | ~3.78e-3        | ~2.3e-5         |
| `decoder_step` | ~3.18e-2        | ~4.8e-5         |

Tighter on every submodel — mostly because there is no opset 17 downgrade pass and FX preserves shared buffers (positional encoding) as initialisers instead of duplicating per layer.

## Risks called out in README
- `torch.export` currently traces with `strict=False`; future PyTorch versions may flip the default and reject some in-wrapper Python conditionals.
- Opset auto-clamp assumes `Pad`-class ops still lack a v17 adapter in onnxscript.
- Conventional `decoder.onnx` path uses a `BaseModelOutput` wrapper internally; if `transformers` changes how the decoder consumes `encoder_outputs`, the wrapper may need updating before re-exporting.

## Test plan
- [x] `mel` + `encoder` end-to-end on CPU with parity check (B=2 padded + B=4 batch-dynamic)
- [x] `decoder_init` + `decoder_step` end-to-end on CPU with parity vs PyTorch reference
- [x] Verified `_consolidate_external_data` no-op detection works on dynamo output
- [x] Verified exported `.data` file size matches raw encoder param count (~7.06 GB fp32 → 7.1 GB exported)
- [ ] Try `--legacy-exporter` end-to-end to confirm fallback path still works
- [ ] Try `--device cuda` and `--dtype float16` on a GPU box
- [ ] Try `--conventional-decoder` end-to-end

Closes #4

### CODE REVIEW

**Comment by christopherthompson81** [comment]
## CUDA test + RTF — RTX 3090

Pushed e4d358d which fixes three CUDA-path constraint failures the CPU trace happened to bypass:

1. Derived dims (e.g. `8*time_frames - 3` from conv subsampling) reject `Dim('name', min=1)` because `min` doesn't propagate through internal symbol rename → switched non-shared dims to `Dim.AUTO`.
2. CUDA SDPA decomposition asserts `batch <= 65535` → added explicit `max=65535` to every batch Dim.
3. CUDA SDPA also specialises B=1 to a static value → bumped decoder_init / decoder_step dummies to B=2 (exported graph still accepts B=1 at runtime).

### Export parity (CUDA, fp32, 5 s dummy)
| Submodel       | CPU max-diff | CUDA max-diff |
| -------------- | ------------ | ------------- |
| `mel`          | 3.5e-5       | 3.6e-5        |
| `encoder`      | 3e-6         | 5.82e-4       |
| `decoder_init` | 2.3e-5       | 3.37e-3       |
| `decoder_step` | 4.8e-5       | 1.43e-2       |

CUDA parity is looser than CPU (different SDPA decompositions accumulate fp differently) but stays within the original legacy reference tolerances from issue #4.

### RTF — 60.03 s audio, greedy decode to natural EOS, 3 runs
```
Audio duration   : 60.032 s
Tokens generated : 212
mel              :   5.4 ms
encoder          : 183.8 ms
decoder_init     :   9.5 ms
decoder_step     : 1771.8 ms over 211 steps  (mean 8.40, p95 9.82 ms/step)
total wall       : 1.974 s
RTF              : 0.0329  (≈ 30× real-time)
```
Decoder dominates wall time (~90%); mel is essentially free; encoder is the only sub-second non-trivial chunk and runs once per utterance.

Two ORT warnings worth flagging (non-blocking):
- `2 Memcpy nodes are added to the graph for CUDAExecutionProvider` — likely the int64 token / position inputs landing on CPU first; could be optimised by binding inputs as CUDA tensors via IO binding.
- `ScatterND with reduction=='none' only guarantees to be correct if indices are not duplicated.` — emitted by the dynamo-side KV concat lowering; indices are non-duplicated in our path, so safe.

**Comment by christopherthompson81** [comment]
## Two more boxes ticked

### `--conventional-decoder` end-to-end (933da8f)
Same B=1 specialisation issue as the KV decoder bit me here too: `make_decoder_dummy_inputs` produced a `(1, 1)` BOS tensor and `main()` sliced the encoder hidden to B=1 before passing it in. CUDA SDPA decomposition then refused to keep batch symbolic. Fix: bump dummy to `(B=2, seq=2)` and pass the full B=2 encoder hidden through.

Result: `decoder.onnx` exports cleanly on CUDA, parity max diff `3.4e-5`, validator passes, weight file 650 MB (matches `transf_decoder` param count).

### C# inference layer end-to-end
Ran the Vernacula CLI against the dynamo CUDA export from earlier (`/tmp/cohere-onnx-cuda`):

```
./Vernacula.CLI --audio demo-audio.wav --model ~/.local/share/Vernacula/models \
  --asr cohere --cohere-model /tmp/cohere-onnx-cuda --segments segs.json --language en
```

Loaded successfully and produced a coherent ~1-minute transcription. The opset 17→18 bump and the new external-data layout (single `<name>.onnx.data` per model, no scattered legacy files) were transparent to the C# loader. The hardcoded KV I/O names (`self_key_{i}`, etc.) and config.json token-ID keys all matched.

Three ORT warnings appeared, all already noted in the previous comment (Memcpy nodes, ScatterND-with-duplicates safe in our path, "shape ops to CPU" cosmetic) — none affected output correctness.

So the only remaining untested path is `--dtype float16`, which the user explicitly opted to skip.
